server/src/main/java/game/ai/BotFactory.java

```diff
@@ -18,4 +19,53 @@ public static Bot createBot(World world) {
 
         return new StupidBot(world);
     }
+
+    /**
+     * Creates one of the various Bot types with the type chosen
+     * depending on the current bot population and the target population such that bot distribution
+     * converges towards the target bot distribution
+     *
+     * @param world
+     * @param game
+     * @return
+     */
+    public static Bot createBotDynamically(World world, Game game) {
+
+        final int scaredBotCount = game.countBotsOfType(ScaredBot.class);
+        final int kamikazeBotCount = game.countBotsOfType(KamikazeBot.class);
+        final int stupidBotCount = game.countBotsOfType(StupidBot.class);
+
+        final int botCount = game.getNumberOfBots();
+        assert (scaredBotCount + kamikazeBotCount + stupidBotCount == botCount);
+
+        if (botCount == 0) {
+            return new StupidBot(world);
+        }
+
+        final double scaredBotRatio = scaredBotCount / (double) botCount;
+        final double kamikazeBotRatio = kamikazeBotCount / (double) botCount;
+        final double stupidBotRatio = stupidBotCount / (double) botCount;
+
+        final double scaredBotRatioDiscrepancy = game.config.botPopulationDistributionTarget.scaredBotRatio() -
+                scaredBotRatio;
+        final double kamikazeBotRatioDiscrepancy = game.config.botPopulationDistributionTarget.kamikazeBotRatio()
+                - kamikazeBotRatio;
+        final double stupidBotRatioDiscrepancy = game.config.botPopulationDistributionTarget.stupidBotRatio() -
+                stupidBotRatio;
```

Comment:
Please store `game.config.botPopulationDistributionTarget` in a shorter variable to make this more readable.